solvers/optimization_utils.cc

```diff
@@ -0,0 +1,175 @@
+#include "solvers/optimization_utils.h"
+
+using Eigen::MatrixXd;
+using Eigen::VectorXd;
+using drake::solvers::Constraint;
+using drake::solvers::Binding;
+using drake::solvers::MathematicalProgram;
+using drake::AutoDiffVecXd;
+using drake::math::initializeAutoDiff;
+using drake::math::autoDiffToGradientMatrix;
+using drake::math::autoDiffToValueMatrix;
+
+namespace dairlib {
+namespace solvers {
+
+void CheckGenericConstraints(const MathematicalProgram& prog,
+    const drake::solvers::MathematicalProgramResult& result,
+    double tol) {
+  for (auto const& binding : prog.generic_constraints()) {
+    auto y = result.EvalBinding(binding);
+    auto c = binding.evaluator();
+    bool isSatisfied = (y.array() >= c->lower_bound().array() - tol).all() &&
+                       (y.array() <= c->upper_bound().array() + tol).all();
+    if (!isSatisfied) {
+      std::cout << "Constraint violation: " <<
+          c->get_description() << std::endl;
+      MatrixXd tmp(y.size(), 3);
+      tmp << c->lower_bound(), y, c->upper_bound();
+      std::cout << tmp << std::endl;
+    }
+  }
+}
+
+double SecondOrderCost(const MathematicalProgram& prog, const VectorXd& x_nom,
+    MatrixXd* Q, VectorXd* w, double eps) {
```

Comment:
Just curious, why do we change from reference to pointer?

Reply:
See the GSG here: https://google.github.io/styleguide/cppguide.html#Reference_Arguments

The main argument is that non-const references are confusing,making it too easy to unwittingly change data. Using pointers and const references makes this more explicit.

Reply:
Got it!